In [1]:
import pandas as pd

In [20]:
df = pd.read_csv(r'C:\Users\vj0467\OneDrive - Sandvik\HomeShare\Documents\Yurchenko desktop copy\Marketing\Нетология_2020\HW\pyda-16-hw-1\horse.csv',  usecols=[0,1,2,3,4,5,6,7], names = ["surgery?", "Age", " Hospital Number", "rectal temperature", "pulse", "respiratory rate", " temperature of extremities", "peripheral pulse"], na_values='?')

In [101]:
df

surgery?  Age   Hospital Number  rectal temperature  pulse  \
0         2.0    1            530101                38.5   66.0   
1         1.0    1            534817                39.2   88.0   
2         2.0    1            530334                38.3   40.0   
3         1.0    9           5290409                39.1  164.0   
4         2.0    1            530255                37.3  104.0   
..        ...  ...               ...                 ...    ...   
295       1.0    1            533886                 NaN  120.0   
296       2.0    1            527702                37.2   72.0   
297       1.0    1            529386                37.5   72.0   
298       1.0    1            530612                36.5  100.0   
299       1.0    1            534618                37.2   40.0   

     respiratory rate   temperature of extremities  peripheral pulse  
0                28.0                          3.0               3.0  
1                20.0                          NaN               NaN  
2                24.0                          1.0               1.0  
3                84.0                          4.0               1.0  
4                35.0                          NaN               NaN  
..                ...                          ...               ...  
295              70.0                          4.0               NaN  
296              24.0                          3.0               2.0  
297              30.0                          4.0               3.0  
298              24.0                          3.0               3.0  
299              20.0                          NaN               NaN  

[300 rows x 8 columns]

In [94]:
print(df.pulse.max())
print(df.pulse.min())
print(df.pulse.mean())
print(df.pulse.mode()[0])
print(df.pulse.median())
print(df.pulse.std())
print(df.pulse.var())
print(df.pulse.quantile([0.25, 0.75]))

184.0
30.0
71.91304347826087
48.0
64.0
28.630556660735003
819.7087747035575
0.25    48.0
0.75    88.0
Name: pulse, dtype: float64


In [95]:
print(df['rectal temperature'].mode()[0])

38.0


In [96]:
df.describe()

surgery?         Age   Hospital Number  rectal temperature  \
count  299.000000  300.000000      3.000000e+02          240.000000   
mean     1.397993    1.640000      1.085889e+06           38.167917   
std      0.490305    2.173972      1.529801e+06            0.732289   
min      1.000000    1.000000      5.184760e+05           35.400000   
25%      1.000000    1.000000      5.289040e+05           37.800000   
50%      1.000000    1.000000      5.303055e+05           38.200000   
75%      2.000000    1.000000      5.347275e+05           38.500000   
max      2.000000    9.000000      5.305629e+06           40.800000   

            pulse  respiratory rate   temperature of extremities  \
count  276.000000        242.000000                   244.000000   
mean    71.913043         30.417355                     2.348361   
std     28.630557         17.642231                     1.045054   
min     30.000000          8.000000                     1.000000   
25%     48.000000         18.500000                     1.000000   
50%     64.000000         24.500000                     3.000000   
75%     88.000000         36.000000                     3.000000   
max    184.000000         96.000000                     4.000000   

       peripheral pulse  
count        231.000000  
mean           2.017316  
std            1.042428  
min            1.000000  
25%            1.000000  
50%            2.000000  
75%            3.000000  
max            4.000000

Сильный разбег в дыхании - параметром можно пренебречь.
В возрасте должно быть всего 2 значения, остальное - ошибки, которые нужно исправить (выбросы заменить).
Температура и пульс будут ключевыми понятиями. В температуре много пропусков, можно заменить значением моды (хотя различие между тремя средними незначительное).
Температура конечностей - качественный показатель, много пропусков. Заменим модой. То же самое касается периферийного пульса.


In [72]:
q1_1 = df.Age.quantile(0.25)
q3_3 = df.Age.quantile(0.75)
iqr_1 = q3_3 - q1_1
lower_bound_1 = q1_1 - (1.5 *iqr_1)
upper_bound_1 = q3_3 + (1.5 *iqr_1)
remove_outliers_age = df[df.Age.between(lower_bound_1, upper_bound_1, inclusive = True)].sort_values('rectal temperature')
remove_outliers_age_1 = remove_outliers_age[['Age', 'rectal temperature']]
remove_outliers_age_1

Age  rectal temperature
44     1                35.4
141    1                36.0
238    1                36.1
80     1                36.4
298    1                36.5
..   ...                 ...
274    1                 NaN
282    1                 NaN
288    1                 NaN
293    1                 NaN
295    1                 NaN

[276 rows x 2 columns]

In [73]:
df_age = df[['Age', 'rectal temperature']]
df_age

Age  rectal temperature
0      1                38.5
1      1                39.2
2      1                38.3
3      9                39.1
4      1                37.3
..   ...                 ...
295    1                 NaN
296    1                37.2
297    1                37.5
298    1                36.5
299    1                37.2

[300 rows x 2 columns]

In [74]:
pd.concat([df_age, remove_outliers_age_1]).drop_duplicates(keep = False)

Age  rectal temperature
3      9                39.1
39     9                39.2
41     9                39.0
75     9                39.7
109    9                38.4
212    9                38.9
229    9                38.5
244    9                38.2
287    9                39.5

выброс в возрасте - точно ошибка. Надо заменить модой и выбросы, и пропуски

In [75]:
age_filled_mode = df['Age'].fillna(df['Age'].mode())
age_filled_mode

0      1
1      1
2      1
3      9
4      1
      ..
295    1
296    1
297    1
298    1
299    1
Name: Age, Length: 300, dtype: int64

заменили пропуски на моду

In [60]:
q1 = df.pulse.quantile(0.25)
q3 = df.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound= q1 - (1.5 *iqr)
upper_bound= q3 + (1.5 *iqr)
remove_outliers = df[df.pulse.between(lower_bound, upper_bound, inclusive = True)].sort_values('rectal temperature')
remove_outliers_1 = remove_outliers[['pulse', 'rectal temperature']]
remove_outliers_1

pulse  rectal temperature
44   140.0                35.4
141   42.0                36.0
238   88.0                36.1
80    98.0                36.4
298  100.0                36.5
..     ...                 ...
272  100.0                 NaN
274   76.0                 NaN
282   70.0                 NaN
293   78.0                 NaN
295  120.0                 NaN

[271 rows x 2 columns]

In [61]:
df_1 = df[['pulse', 'rectal temperature']]
df_1


pulse  rectal temperature
0     66.0                38.5
1     88.0                39.2
2     40.0                38.3
3    164.0                39.1
4    104.0                37.3
..     ...                 ...
295  120.0                 NaN
296   72.0                37.2
297   72.0                37.5
298  100.0                36.5
299   40.0                37.2

[300 rows x 2 columns]

In [64]:
pd.concat([df_1, remove_outliers_1]).drop_duplicates(keep = False)

pulse  rectal temperature
3    164.0                39.1
41   150.0                39.0
55   160.0                38.6
117    NaN                39.5
126    NaN                38.7
151    NaN                39.3
204    NaN                39.2
255  184.0                38.8
275  150.0                38.8

Нашли выбросы для пульса. Такие значение пульса (160, 184) возможны, если лошадь в движении. Здесь лошади вряд ли бегали во время измерений, поэтому такое повышение пульса маловероятно. Эти выбросы лучше заменить на среднее арифметич  

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery?                     299 non-null    float64
 1   Age                          300 non-null    int64  
 2    Hospital Number             300 non-null    int64  
 3   rectal temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   respiratory rate             242 non-null    float64
 6    temperature of extremities  244 non-null    float64
 7   peripheral pulse             231 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 18.9 KB


In [54]:
for col in df.columns:
    pct_missing = df[col].isna().mean()
    print (f'{col} - {pct_missing :.1%}')

surgery? - 0.3%
Age - 0.0%
 Hospital Number - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
respiratory rate - 19.3%
 temperature of extremities - 18.7%
peripheral pulse - 23.0%


заменим проуски для важных колонок (пульс, температура)

In [55]:
rectal_temperature_filled_mean = df['rectal temperature'].fillna(df['rectal temperature'].mean())
rectal_temperature_filled_mean

0      38.500000
1      39.200000
2      38.300000
3      39.100000
4      37.300000
         ...    
295    38.167917
296    37.200000
297    37.500000
298    36.500000
299    37.200000
Name: rectal temperature, Length: 300, dtype: float64

In [58]:

print(rectal_temperature_filled_mean.mean())
print(rectal_temperature_filled_mean.mode()[0])
print(rectal_temperature_filled_mean.median())
print(rectal_temperature_filled_mean.std())
print(rectal_temperature_filled_mean.var())


38.16791666666669
38.16791666666669
38.16791666666669
0.6547050149748525
0.42863865663322176


около 19% значений температуры были пропущены. Это важный показатель, поэтому его нужно сохранить. Заменим на среднее арифметическое, хотя изначальные значения моды, медианы и ср арифметич были очень близки (можно было выбрать любое).

частота дыхания также имеет большое количество пропусков, но колебания этого показателя достаточно сильные, а ценность не высокая. 

In [70]:
pulse_filled_median = df['pulse'].fillna(df['pulse'].median())
pulse_filled_median

0       66.0
1       88.0
2       40.0
3      164.0
4      104.0
       ...  
295    120.0
296     72.0
297     72.0
298    100.0
299     40.0
Name: pulse, Length: 300, dtype: float64

In [71]:

print(pulse_filled_median.mean())
print(pulse_filled_median.mode()[0])
print(pulse_filled_median.median())
print(pulse_filled_median.std())
print(pulse_filled_median.var())

71.28
64.0
64.0
27.54154538888526
758.536722408027


Для пульса среднее арифметичсекое было сильно выше моды, поэтому это не самый удачный вариант для заполнения пропусков, их 8%. Возьмем медиану, тогда значения моды и медианы совпадут, а ср арифметическое станет немного ниже.

In [121]:
# заменим модой пропуски для температуры конечностей и перф пульса
p_pulse_filled_mode = df['peripheral pulse'].fillna(df['peripheral pulse'].mode()[0])
p_pulse_filled_mode 

0      3.0
1      1.0
2      1.0
3      1.0
4      1.0
      ... 
295    1.0
296    2.0
297    3.0
298    3.0
299    1.0
Name: peripheral pulse, Length: 300, dtype: float64

In [110]:
temp_extrem_mode = df[' temperature of extremities'].fillna(df[' temperature of extremities'].mode()[0])
temp_extrem_mode

0      3.0
1      NaN
2      1.0
3      4.0
4      NaN
      ... 
295    4.0
296    3.0
297    4.0
298    3.0
299    NaN
Name:  temperature of extremities, Length: 300, dtype: float64

заменили пропуски для периф пульса и температуры конечностей модой